rs/index_writer/src/input/hdf5.rs

```diff
@@ -0,0 +1,72 @@
+use std::cmp::min;
+
+use anyhow::Result;
+use ndarray::{s, Array2};
+
+pub struct Hdf5Reader {
+    dataset: hdf5::Dataset,
+    chunk_size: usize,
+    row_idx: usize,
+    num_rows: usize,
+    chunk: Vec<Vec<f32>>,
+}
+
+impl Hdf5Reader {
+    pub fn new(chunk_size: usize, dataset: &str, file: &str) -> Result<Self> {
+        let file = hdf5::File::open(file)?;
+        let dataset = file.dataset(dataset)?;
+        let num_rows = dataset.shape()[0];
+        Ok(Self {
+            dataset,
+            chunk_size,
+            row_idx: 0,
+            num_rows,
+            chunk: vec![],
+        })
+    }
+
+    pub fn has_next(&self) -> bool {
+        self.row_idx < self.num_rows
+    }
+
+    // Caller is responsible for ensuring that the chunk is not empty
+    pub fn next(&mut self) -> &[f32] {
+        if self.row_idx % self.chunk_size == 0 {
+            self.fetch_next_chunk();
+        }
+
+        let idx = self.row_idx % self.chunk_size;
+        let slice = &self.chunk[idx];
+        self.row_idx += 1;
+        slice
+    }
+
+    pub fn fetch_next_chunk(&mut self) {
```

Comment:
This should be private? Letting external code calls this function will mess up `row_idx`.